cloud/terraform/terraform_controller.py

```diff
@@ -0,0 +1,122 @@
+                print("Before adding any aws region, aws_credentials should be called")
+                exit()
+
+            region_block = """
+provider "aws" {{
+  alias   = "{region}"
+  region  = "{region}"
+  access_key = "{access_key}"
+  secret_key = "{secret_key}"
+  token = "{token}"
+}}
+""".format(region=region, access_key=self.aws_access_key, secret_key=self.aws_secret_key, token=self.token)
+            self._add_to_config_file(region_block)
+        
+    def add_instance(self, region, image_id, instance_type=""):
+        """
+        Add instance to configuration and returns the terraform resource id
+        """
+        if not self._string_in_config(region):
+            print("Cannot add an instance if region provider is not set up")
+            exit()
+        
+        if self.cloud == "aws":
+            if not instance_type:
+                instance_type = "t2.micro"
+            instance_block ="""
+resource "aws_instance" "instance-{region}" {{
+  provider      = aws.{region}
+  ami           = "{image_id}"
+  instance_type = "{instance_type}"
+  tags = {{
+    Name = "instance-{region}"
+  }}
+}}           
+""".format(region=region, image_id=image_id, instance_type=instance_type)
+            self._add_to_config_file(instance_block)
+            return f"aws_instance.instance-{region}"
+
+    def create_infra(self):
+        """
+        Create the infrastucture specified in the terraform configuration
+        """
+        cmd_output = os.system("terraform init")
+        if cmd_output:
+            print("terraform init command failed, check configuration")
+            exit()
+
+        cmd_output = os.system("terraform apply")
+        if cmd_output:
+            print("terraform apply command failed, check configuration")
+            exit()
+
+    def destroy_resource(self, resource_id):
+        """
+        Destroys specific resource
+        """
+        cmd_output = os.system(f"terraform destroy -target={resource_id}")
+        if cmd_output:
+            print("terraform destroy specific resource command failed")
+            exit()
+
+    def destroy_infra(self):
+        cmd_output = os.system("terraform destroy -auto-approve")
+        if cmd_output:
+            print("terraform destroy command failed")
+            exit()
+
+if __name__ == "__main__":
+    test_controller = TerraformController("aws")
+    # Use temp credentials while developing
+    test_controller.aws_credentials("ASIAWMIV2JB6AUQBMWOV", "hacqjoJyEpQ8gRNajiQNtubL/kmYjIQtaS1sqQLV", "FwoGZXIvYXdzEKH//////////wEaDIQgocVdYvRAGcVjESLmAZ88WUi1DZHkfMuM+y1Dyt6t3H15R59+6RqsJLuf7WB/MaK0OgAlMAzEWUGC2mBYtafWUh7zabiAkCxv0HDXfBwSWnCW0wVvRAI7RM4wcLUckhUF/xwtMhJIOuCdj2/IfgyqFs3a3v5MqxP+GOe6kSjh79zT2X7GnegE+T+uE9hSHddlZsVIdRZp1ulwhiY5mnxhDgh8q9up2+yO0MHveDKOuH69vLlDhRgL1otTWtMNw4mtndDwyaFIPwyJbiNumoz9zFUc5EtUq1/CEQH/sDzScKWAy1e6tqwKFQvTiC9AFlnyYjKHKK674pEGMjJm4OrM4fxo0VaaxX+O5VcdxPH2dYZ6tpoki0nFUsJUNAsqSJzG+proXv/Y/6OQ/SJ7KA==")
```

Comment:
> keys for testing and may have expired already

The `may` scares me. Please don't push any AWS credentials to a repo, it's easy to mess up and push live ones.

Reply:
They expired when I pushed it. But yeah, I'll use credential files from now on